game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1221 @@
+      return getRegularFightOutcomes(aliveOrInjuredUnits, Matches.unitIsFirstStrike(), returnFire);
+    } else if (aliveOrInjuredUnits.getType() == Type.SUB_DEFENDER) {
+      final MustFightBattle.ReturnFire returnFire =
+          returnFireAgainstAttackingSubs(
+              units.getAliveOrWaitingToDieEnemy(), units.getAliveOrWaitingToDieFriendly());
+      return getRegularFightOutcomes(aliveOrInjuredUnits, Matches.unitIsFirstStrike(), returnFire);
+    } else {
+      return getRegularFightOutcomes(
+          aliveOrInjuredUnits,
+          Matches.unitIsFirstStrike().negate(),
+          MustFightBattle.ReturnFire.ALL);
+    }
+  }
+
+  private List<StepUnits> getRegularFightOutcomes(
+      final StepUnits aliveOrInjuredUnits,
+      final Predicate<Unit> firingUnitPredicate,
+      final MustFightBattle.ReturnFire returnFire) {
+    final Collection<Unit> allFiringUnits =
+        CollectionUtils.getMatches(
+            aliveOrInjuredUnits.getAliveOrWaitingToDieFriendly(), firingUnitPredicate);
+    final Collection<Unit> allEnemyUnits = aliveOrInjuredUnits.getAliveOrWaitingToDieEnemy();
+    final List<Triple<Collection<Unit>, Collection<Unit>, Boolean>> groupsAndTargets =
+        new ArrayList<>();
+    final List<TargetGroup> targetGroups =
+        TargetGroup.newTargetGroups(allFiringUnits, allEnemyUnits);
+    for (final TargetGroup targetGroup : targetGroups) {
+      final Collection<Unit> firingUnits = targetGroup.getFiringUnits(allFiringUnits);
+      final Collection<Unit> attackableUnits = targetGroup.getTargetUnits(allEnemyUnits);
+      final Collection<Unit> targetUnits =
+          CollectionUtils.getMatches(
+              attackableUnits,
+              PredicateBuilder.of(Matches.unitIsNotInfrastructure())
+                  .andIf(
+                      aliveOrInjuredUnits.getType() == Type.DEFENDER,
+                      Matches.unitIsSuicideOnAttack().negate())
+                  .andIf(
+                      aliveOrInjuredUnits.getType() == Type.ATTACKER,
+                      Matches.unitIsSuicideOnDefense().negate())
+                  .build());
+
+      if (targetUnits.isEmpty() || firingUnits.isEmpty()) {
+        continue;
+      }
+      final List<Collection<Unit>> firingGroups = MustFightBattle.newFiringUnitGroups(firingUnits);
+      for (final Collection<Unit> firingGroup : firingGroups) {
+        final Boolean isSuicideOnHit = firingGroup.stream().anyMatch(Matches.unitIsSuicideOnHit());
+        groupsAndTargets.add(Triple.of(firingGroup, targetUnits, isSuicideOnHit));
+      }
+    }
+    final Map<StepUnits, Double> outcomes = new HashMap<>();
+
+    if (!groupsAndTargets.isEmpty()) {
+      getTargetGroupFightOutcomes(
+          outcomes,
+          groupsAndTargets,
+          groupsAndTargets.size() - 1,
+          aliveOrInjuredUnits.swapSides(),
+          aliveOrInjuredUnits.getType(),
+          returnFire,
+          true,
+          1.0);
+    } else {
```

Comment:
If you invert the if and just return, then the code in the above block could just be moved out (no need for else).